examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,39 +18,152 @@
+    VerifyOrExit(messageIdBuffer != nullptr,
+                 ChipLogProgress(Controller, "HandlePresentMessagesRequest messageIdBuffer alloc failed"));
+    memcpy(messageIdBuffer, messageId.data(), messageId.size());
+
+    if (responses.HasValue())
+    {
+        size_t size    = 0;
+        CHIP_ERROR err = responses.Value().ComputeSize(&size);
+        VerifyOrExit(err == CHIP_NO_ERROR, ChipLogProgress(Controller, "HandlePresentMessagesRequest size check failed"));
+
+        MessageResponseOption * optionArray = new MessageResponseOption[size];
+        VerifyOrExit(optionArray != nullptr,
+                     ChipLogProgress(Controller, "HandlePresentMessagesRequest MessageResponseOption alloc failed"));
+        optionToFree.push_back(optionArray);
+
+        int counter = 0;
+        auto iter   = responses.Value().begin();
+        while (iter.Next())
+        {
+            auto & response = iter.GetValue();
+
+            ChipLogProgress(Controller, "HandlePresentMessagesRequest option id:%u", response.messageResponseID.ValueOr(0));
+
+            if (response.messageResponseID.HasValue())
+            {
+                optionArray[counter].messageResponseID = Optional<uint32_t>(response.messageResponseID.ValueOr(0));
+            }
+
+            if (response.label.HasValue())
+            {
+                char * labelBuffer = new char[response.label.Value().size()];
+                VerifyOrExit(labelBuffer != nullptr,
+                             ChipLogProgress(Controller, "HandlePresentMessagesRequest label alloc failed"));
+                needToFree.push_back(labelBuffer);
+                memcpy(labelBuffer, response.label.Value().data(), response.label.Value().size());
+
+                ChipLogProgress(Controller, "HandlePresentMessagesRequest option label:%s size:%lu", labelBuffer,
+                                response.label.Value().size());
+
+                optionArray[counter].label = Optional<CharSpan>(CharSpan::fromCharString(labelBuffer));
+            }
+            counter++;
+        }
+
+        Message message{ ByteSpan(messageIdBuffer, messageId.size()),
+                         priority,
+                         messageControl,
+                         startTime,
+                         duration,
+                         CharSpan::fromCharString(messageTextBuffer),
+                         Optional<DataModel::List<MessageResponseOption>>(
+                             DataModel::List<MessageResponseOption>(optionArray, size)) };
+
+        mMessages.push_back(message);
+    }
+    else
+    {
+        Message message{ ByteSpan(messageIdBuffer, messageId.size()), priority, messageControl, startTime, duration,
+                         CharSpan::fromCharString(messageTextBuffer) };
+        mMessages.push_back(message);
+    }
+
     // Add your code to present Message
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest complete");
+
+    return;
```

Comment:
where is the freeing of the allocated things if we return here? I see some freeing in cancel ... is cancel guaranteed to return?

Memory management here is not very clear, and if not very clear it is hard to maintain (or even make sure it is bug-free)

Reply:
refactored the freeing to the new CacheMessage struct